In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/earthquake-database/database.csv


### KAGGLE HAQIDA

KAGGLE PLATFORMASI DATA SCIENCE mutaxasislari uchun eng katta onlayn platforma hisoblanadi.<br>
Bu platforamda imkoniyatlar juda kop. Jupiter Notebukingizni mana shuni ozida ishga tushurishingiz mumkin. <br>
KAGGLEda millionlab datasetlar jamlangan. Istalgan mavzuda bu yerdan dataset topishingiz mumkin. <br>
Shuningdek bu yerda turli xil sovrinli musobaqalar otkazib turiladi. Bu yerda jamoangiz yoki <br>
ozingiz mustaqil qatnashib korishingiz mumkin. Hatto bepul kurslargacha mavjud

In [3]:
df = pd.read_csv("/kaggle/input/earthquake-database/database.csv")
df

Date      Time  Latitude  Longitude        Type   Depth  \
0      01/02/1965  13:44:18   19.2460   145.6160  Earthquake  131.60   
1      01/04/1965  11:29:49    1.8630   127.3520  Earthquake   80.00   
2      01/05/1965  18:05:58  -20.5790  -173.9720  Earthquake   20.00   
3      01/08/1965  18:49:43  -59.0760   -23.5570  Earthquake   15.00   
4      01/09/1965  13:32:50   11.9380   126.4270  Earthquake   15.00   
...           ...       ...       ...        ...         ...     ...   
23407  12/28/2016  08:22:12   38.3917  -118.8941  Earthquake   12.30   
23408  12/28/2016  09:13:47   38.3777  -118.8957  Earthquake    8.80   
23409  12/28/2016  12:38:51   36.9179   140.4262  Earthquake   10.00   
23410  12/29/2016  22:30:19   -9.0283   118.6639  Earthquake   79.00   
23411  12/30/2016  20:08:28   37.3973   141.4103  Earthquake   11.94   

       Depth Error  Depth Seismic Stations  Magnitude Magnitude Type  ...  \
0              NaN                     NaN        6.0             MW  ...   
1              NaN                     NaN        5.8             MW  ...   
2              NaN                     NaN        6.2             MW  ...   
3              NaN                     NaN        5.8             MW  ...   
4              NaN                     NaN        5.8             MW  ...   
...            ...                     ...        ...            ...  ...   
23407          1.2                    40.0        5.6             ML  ...   
23408          2.0                    33.0        5.5             ML  ...   
23409          1.8                     NaN        5.9            MWW  ...   
23410          1.8                     NaN        6.3            MWW  ...   
23411          2.2                     NaN        5.5             MB  ...   

       Magnitude Seismic Stations  Azimuthal Gap  Horizontal Distance  \
0                             NaN            NaN                  NaN   
1                             NaN            NaN                  NaN   
2                             NaN            NaN                  NaN   
3                             NaN            NaN                  NaN   
4                             NaN            NaN                  NaN   
...                           ...            ...                  ...   
23407                        18.0          42.47                0.120   
23408                        18.0          48.58                0.129   
23409                         NaN          91.00                0.992   
23410                         NaN          26.00                3.553   
23411                       428.0          97.00                0.681   

       Horizontal Error  Root Mean Square            ID  Source  \
0                   NaN               NaN  ISCGEM860706  ISCGEM   
1                   NaN               NaN  ISCGEM860737  ISCGEM   
2                   NaN               NaN  ISCGEM860762  ISCGEM   
3                   NaN               NaN  ISCGEM860856  ISCGEM   
4                   NaN               NaN  ISCGEM860890  ISCGEM   
...                 ...               ...           ...     ...   
23407               NaN            0.1898    NN00570710      NN   
23408               NaN            0.2187    NN00570744      NN   
23409               4.8            1.5200    US10007NAF      US   
23410               6.0            1.4300    US10007NL0      US   
23411               4.5            0.9100    US10007NTD      US   

      Location Source Magnitude Source     Status  
0              ISCGEM           ISCGEM  Automatic  
1              ISCGEM           ISCGEM  Automatic  
2              ISCGEM           ISCGEM  Automatic  
3              ISCGEM           ISCGEM  Automatic  
4              ISCGEM           ISCGEM  Automatic  
...               ...              ...        ...  
23407              NN               NN   Reviewed  
23408              NN               NN   Reviewed  
23409              US               US   Reviewed  
23410              US

### Matnni DATETIME ga otkazish

Biz jadvalimizdagi sanalar va vaqtlar bilan ham ishlashimizga togri keladi.<br>
VAQT VA SANALAR ni qanday ajratib olish, qanday ishlash, ishlov berishni organamiz:

Pythondagi datetime degan modul bor, bevosita vaqtlar bilan ishlaydigan:

In [7]:
import datetime as dt

# python fundament kurslarimizda bu haqida atroflicha toxtalganmiz

In [8]:
# hozir chiqarmoqchi bolganimizda

dt.datetime.now()

datetime.datetime(2022, 11, 1, 18, 19, 26, 25271)

dataframe haqida statistik malumot olaylik:

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23412 entries, 0 to 23411
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Date                        23412 non-null  object 
 1   Time                        23412 non-null  object 
 2   Latitude                    23412 non-null  float64
 3   Longitude                   23412 non-null  float64
 4   Type                        23412 non-null  object 
 5   Depth                       23412 non-null  float64
 6   Depth Error                 4461 non-null   float64
 7   Depth Seismic Stations      7097 non-null   float64
 8   Magnitude                   23412 non-null  float64
 9   Magnitude Type              23409 non-null  object 
 10  Magnitude Error             327 non-null    float64
 11  Magnitude Seismic Stations  2564 non-null   float64
 12  Azimuthal Gap               7299 non-null   float64
 13  Horizontal Distance         160

Koradigan bolsak bu yerda DATA va TIME object deb korsatilgan yani string malumot turi<br>
Bunday holatda yuqoridagi datetime modulidagi imkoniyatlardan foydalana olmaymiz

Buni oldini olish uchun type: objectni type: datetime ga otkazishimiz kerak:<br>
Pandas bu uchun to_datetime() degan metod bor:



In [11]:
df.at[0, 'Date']

'01/02/1965'

'01/02/1965' bu stringni Datetimega otkazishdan oldin <br>
uni formatini ham korsatish kerak, dataframe yaxshilab qarasak<br>
01 - bu oy, 02 - bu kun, 1956 - yil sifatida kelyapti

oyni ifodalash uchun '%m'<br>
kunni ifodalash uchun '%d'<br>
yil 4talik yo iktalik keladi:<br>
4xonali yillarni ifodalash uchun '%Y'<br>
2xonali yillarni ifodalash uchun '%y'<br>


Biz bu malumotlarni Python strftime cheatsheatdan olyapmiz:<br>
https://strftime.org/

In [15]:
# pd.to_datetime(df['Date'], format='%m/%d/%Y')

agar biz shuni ishga tushirsak xato chiqaradi<br>
`ValueError: time data '1975-02-23T02:58:41.000Z' does not match format '%m/%d/%Y' (match)` <br>

nega xato chiqardi agar dfni koradigan bolsak unda 23412 ta qator bor<br>
aynan Date ustuni qaysidir qatorida yuqoridagi formatdagi malumot yozib qolingan

Demak aynan shunday yozilgan qatorni topib tuzatishimiz yo ozgartirish kiritishimiz kerak<br>
Bu yerda universal yechim yoq, hamma holat uchun alohida yondashish kerak:

In [33]:
df[df['Date'].str.startswith('19')]

Date                      Time  Latitude  Longitude  \
3378  1975-02-23T02:58:41.000Z  1975-02-23T02:58:41.000Z     8.017    124.075   
7512  1985-04-28T02:53:41.530Z  1985-04-28T02:53:41.530Z   -32.998    -71.766   

            Type  Depth  Depth Error  Depth Seismic Stations  Magnitude  \
3378  Earthquake  623.0          NaN                     NaN        5.6   
7512  Earthquake   33.0          NaN                     NaN        5.6   

     Magnitude Type  ...  Magnitude Seismic Stations  Azimuthal Gap  \
3378             MB  ...                         NaN            NaN   
7512             MW  ...                         NaN            NaN   

      Horizontal Distance  Horizontal Error  Root Mean Square          ID  \
3378                  NaN               NaN               NaN  USP0000A09   
7512                  NaN               NaN               1.3  USP0002E81   

     Source Location Source Magnitude Source    Status  
3378     US              US               US  Reviewed  
7512     US              US              HRV  Reviewed  

[2 rows x 21 columns]

In [36]:
df.at[3378, 'Date']

'1975-02-23T02:58:41.000Z'

In [38]:
df.at[3378, 'Date'] = '02/23/1975'

In [39]:
df.at[3378, 'Date']

'02/23/1975'

In [40]:
df.at[3378, 'Time'] = '02:58:41'

In [41]:
df.at[3378, 'Time']

'02:58:41'

In [42]:
df.at[7512, 'Date']

'1985-04-28T02:53:41.530Z'

In [43]:
df.at[7512, 'Date'] = '04/28/1985'
df.at[7512, 'Time'] = '02:53:41'

In [44]:
df.at[7512, 'Date']

'04/28/1985'

In [45]:
df.at[7512, 'Time']

'02:53:41'

Yuqoridagi birinchi usul edi<br>
Lekin bu usulning kamchiligi boshlanishi 1975 dan boshlanmagan notogri kiritilgan vaqtlarni inobatga olmaydi

2-usuli uzunligidan foydalansak boladi
Yani Date ustunimiz aksariyatini uzunligi bir xil faqatgina xato bergandan boshqa

In [46]:
df.Date.str.len()

0        10
1        10
2        10
3        10
4        10
         ..
23407    10
23408    10
23409    10
23410    10
23411    10
Name: Date, Length: 23412, dtype: int64

In [47]:
df.Date.str.len().value_counts()

10    23411
24        1
Name: Date, dtype: int64

Etibor bersak biz 2ta notogri formatni tuzatgan edik,<br>
va yana bir notogri format qolib ketibdi:

Ularni ajratib olish uchun osongina filtrlaymiz:

In [48]:
df[df.Date.str.len() == 24]

Date                      Time  Latitude  \
20650  2011-03-13T02:23:34.520Z  2011-03-13T02:23:34.520Z    36.344   

       Longitude        Type  Depth  Depth Error  Depth Seismic Stations  \
20650    142.344  Earthquake   10.1         13.9                   289.0   

       Magnitude Magnitude Type  ...  Magnitude Seismic Stations  \
20650        5.8            MWC  ...                         NaN   

       Azimuthal Gap  Horizontal Distance  Horizontal Error  Root Mean Square  \
20650           32.3                  NaN               NaN              1.06   

               ID Source Location Source Magnitude Source    Status  
20650  USP000HWQP     US              US             GCMT  Reviewed  

[1 rows x 21 columns]

In [49]:
# yuqoridagi usul bilan buni ham ozgartirib qoyamiz:

df.at[20650, 'Date'] = '03/13/2011'
df.at[7512, 'Time'] = '02:23:34'

In [69]:
df.Date.str.len().value_counts()

10    23412
Name: Date, dtype: int64

Endi biz yuqoridagi qiymatni qayta ishga tushurishimiz mumkin:

In [51]:
pd.to_datetime(df['Date'], format='%m/%d/%Y')

0       1965-01-02
1       1965-01-04
2       1965-01-05
3       1965-01-08
4       1965-01-09
           ...    
23407   2016-12-28
23408   2016-12-28
23409   2016-12-28
23410   2016-12-29
23411   2016-12-30
Name: Date, Length: 23412, dtype: datetime64[ns]

Mana string typedagi datalarni, datetime typedagi datalarga otkazib oldik

Buni yangi ustunga yozib qoysak ham boladi

In [52]:
df['datetime'] = pd.to_datetime(df['Date'], format='%m/%d/%Y')
df

Date      Time  Latitude  Longitude        Type   Depth  \
0      01/02/1965  13:44:18   19.2460   145.6160  Earthquake  131.60   
1      01/04/1965  11:29:49    1.8630   127.3520  Earthquake   80.00   
2      01/05/1965  18:05:58  -20.5790  -173.9720  Earthquake   20.00   
3      01/08/1965  18:49:43  -59.0760   -23.5570  Earthquake   15.00   
4      01/09/1965  13:32:50   11.9380   126.4270  Earthquake   15.00   
...           ...       ...       ...        ...         ...     ...   
23407  12/28/2016  08:22:12   38.3917  -118.8941  Earthquake   12.30   
23408  12/28/2016  09:13:47   38.3777  -118.8957  Earthquake    8.80   
23409  12/28/2016  12:38:51   36.9179   140.4262  Earthquake   10.00   
23410  12/29/2016  22:30:19   -9.0283   118.6639  Earthquake   79.00   
23411  12/30/2016  20:08:28   37.3973   141.4103  Earthquake   11.94   

       Depth Error  Depth Seismic Stations  Magnitude Magnitude Type  ...  \
0              NaN                     NaN        6.0             MW  ...   
1              NaN                     NaN        5.8             MW  ...   
2              NaN                     NaN        6.2             MW  ...   
3              NaN                     NaN        5.8             MW  ...   
4              NaN                     NaN        5.8             MW  ...   
...            ...                     ...        ...            ...  ...   
23407          1.2                    40.0        5.6             ML  ...   
23408          2.0                    33.0        5.5             ML  ...   
23409          1.8                     NaN        5.9            MWW  ...   
23410          1.8                     NaN        6.3            MWW  ...   
23411          2.2                     NaN        5.5             MB  ...   

       Azimuthal Gap  Horizontal Distance  Horizontal Error  Root Mean Square  \
0                NaN                  NaN               NaN               NaN   
1                NaN                  NaN               NaN               NaN   
2                NaN                  NaN               NaN               NaN   
3                NaN                  NaN               NaN               NaN   
4                NaN                  NaN               NaN               NaN   
...              ...                  ...               ...               ...   
23407          42.47                0.120               NaN            0.1898   
23408          48.58                0.129               NaN            0.2187   
23409          91.00                0.992               4.8            1.5200   
23410          26.00                3.553               6.0            1.4300   
23411          97.00                0.681               4.5            0.9100   

                 ID  Source Location Source Magnitude Source     Status  \
0      ISCGEM860706  ISCGEM          ISCGEM           ISCGEM  Automatic   
1      ISCGEM860737  ISCGEM          ISCGEM           ISCGEM  Automatic   
2      ISCGEM860762  ISCGEM          ISCGEM           ISCGEM  Automatic   
3      ISCGEM860856  ISCGEM          ISCGEM           ISCGEM  Automatic   
4      ISCGEM860890  ISCGEM          ISCGEM           ISCGEM  Automatic   
...             ...     ...             ...              ...        ...   
23407    NN00570710      NN              NN               NN   Reviewed   
23408    NN00570744      NN              NN               NN   Reviewed   
23409    US10007NAF      US              US               US   Reviewed   
23410    US10007NL0      US              US               US   Reviewed   
23411    US10007NTD      US              US               US   Reviewed   

        datetime  
0     1965-01-02  
1     1965-01-04  
2     1965-01-05  
3     1965-01-08  
4     1965-01-09  
...          ...  
23407 2016-12-28  
23408 2016-12-28  
23409 2016-12-28  
23410 2016-12-29  
23411 2016-12-30  

[23412 rows x 22 columns]

### datetime bilan ishlash

Datetime tipiga otkazib olgandan keyin unda ancha ishlar qilish mumkin

datetima ustunimizni 1-malumotini chiqarib olaylik:

In [54]:
df.at[0, 'datetime']

Timestamp('1965-01-02 00:00:00')

In [56]:
# bundan oyni ajratib olishimiz mumkin

df.at[0, 'datetime'].month

1

In [58]:
# haftani qaysi kuni, 5-indeks Shanba kuni, 0 dan boshlanadi

df.at[0, 'datetime'].weekday()

5

In [62]:
# oyni alohida ajratib olishimiz mumkin

df['month'] = df['datetime'].dt.month

# bu degani df jadvalidan datetime ustuni olib undan faqat oylarni ajratib ol degani
# va ajratib olingan oylarni df.month yangi ustuniga yuklab chiq degani

In [63]:
df.month.value_counts()

3     2114
8     2014
12    2001
11    1987
9     1985
4     1971
5     1964
10    1952
1     1891
7     1880
2     1829
6     1824
Name: month, dtype: int64

Demak eng kop zilzilar (2114ta) 3-oy yani Mart oyida bolgan ekan

In [64]:
# hafta kunlarini alohida ajratib olishimiz mumkin

df['weekday'] = df['datetime'].dt.weekday

In [65]:
df.weekday.value_counts()

5    3433
2    3431
4    3362
0    3349
3    3306
6    3287
1    3244
Name: weekday, dtype: int64

Demak eng kop zilzilar (3433ta) 5-weekday yani Shanba kunida bolgan ekan

Time degan ustunni ham datetimega otkazib olishimiz mumkin

In [70]:
# df['time'] = pd.to_datetime(df.Time, format='%H:%M:%S')
# ishga tushursak xato berdi
# ValueError: time data '2011-03-13T02:23:34.520Z' does not match format '%H:%M:%S' (match)

# Endi bunga ham yuqoridagi kabi ishlov bersak boladi

In [71]:
df.Time.str.len().value_counts()

8     23411
24        1
Name: Time, dtype: int64

In [72]:
df[df.Time.str.len() == 24]

Date                      Time  Latitude  Longitude        Type  \
20650  03/13/2011  2011-03-13T02:23:34.520Z    36.344    142.344  Earthquake   

       Depth  Depth Error  Depth Seismic Stations  Magnitude Magnitude Type  \
20650   10.1         13.9                   289.0        5.8            MWC   

       ...  Horizontal Error  Root Mean Square          ID  Source  \
20650  ...               NaN              1.06  USP000HWQP      US   

       Location Source  Magnitude Source    Status   datetime month weekday  
20650               US              GCMT  Reviewed 2011-03-13     3       6  

[1 rows x 24 columns]

In [73]:
df.loc[20650, 'Time'] = '02:23:34'
df.loc[20650, 'Time']

'02:23:34'

In [74]:
df.Time.str.len().value_counts()

8    23412
Name: Time, dtype: int64

In [75]:
# xatolikni tuzatdik endi formatga otkazsak boladi

df['time'] = pd.to_datetime(df.Time, format='%H:%M:%S')

In [76]:
df

Date      Time  Latitude  Longitude        Type   Depth  \
0      01/02/1965  13:44:18   19.2460   145.6160  Earthquake  131.60   
1      01/04/1965  11:29:49    1.8630   127.3520  Earthquake   80.00   
2      01/05/1965  18:05:58  -20.5790  -173.9720  Earthquake   20.00   
3      01/08/1965  18:49:43  -59.0760   -23.5570  Earthquake   15.00   
4      01/09/1965  13:32:50   11.9380   126.4270  Earthquake   15.00   
...           ...       ...       ...        ...         ...     ...   
23407  12/28/2016  08:22:12   38.3917  -118.8941  Earthquake   12.30   
23408  12/28/2016  09:13:47   38.3777  -118.8957  Earthquake    8.80   
23409  12/28/2016  12:38:51   36.9179   140.4262  Earthquake   10.00   
23410  12/29/2016  22:30:19   -9.0283   118.6639  Earthquake   79.00   
23411  12/30/2016  20:08:28   37.3973   141.4103  Earthquake   11.94   

       Depth Error  Depth Seismic Stations  Magnitude Magnitude Type  ...  \
0              NaN                     NaN        6.0             MW  ...   
1              NaN                     NaN        5.8             MW  ...   
2              NaN                     NaN        6.2             MW  ...   
3              NaN                     NaN        5.8             MW  ...   
4              NaN                     NaN        5.8             MW  ...   
...            ...                     ...        ...            ...  ...   
23407          1.2                    40.0        5.6             ML  ...   
23408          2.0                    33.0        5.5             ML  ...   
23409          1.8                     NaN        5.9            MWW  ...   
23410          1.8                     NaN        6.3            MWW  ...   
23411          2.2                     NaN        5.5             MB  ...   

       Root Mean Square            ID  Source  Location Source  \
0                   NaN  ISCGEM860706  ISCGEM           ISCGEM   
1                   NaN  ISCGEM860737  ISCGEM           ISCGEM   
2                   NaN  ISCGEM860762  ISCGEM           ISCGEM   
3                   NaN  ISCGEM860856  ISCGEM           ISCGEM   
4                   NaN  ISCGEM860890  ISCGEM           ISCGEM   
...                 ...           ...     ...              ...   
23407            0.1898    NN00570710      NN               NN   
23408            0.2187    NN00570744      NN               NN   
23409            1.5200    US10007NAF      US               US   
23410            1.4300    US10007NL0      US               US   
23411            0.9100    US10007NTD      US               US   

       Magnitude Source     Status   datetime month weekday  \
0                ISCGEM  Automatic 1965-01-02     1       5   
1                ISCGEM  Automatic 1965-01-04     1       0   
2                ISCGEM  Automatic 1965-01-05     1       1   
3                ISCGEM  Automatic 1965-01-08     1       4   
4                ISCGEM  Automatic 1965-01-09     1       5   
...                 ...        ...        ...   ...     ...   
23407                NN   Reviewed 2016-12-28    12       2   
23408                NN   Reviewed 2016-12-28    12       2   
23409                US   Reviewed 2016-12-28    12       2   
23410                US   Reviewed 2016-12-29    12       3   
23411                US   Reviewed 2016-12-30    12       4   

                     time  
0     1900-01-01 13:44:18  
1     1900-01-01 11:29:49  
2     1900-01-01 18:05:58  
3     1900-01-01 18:49:43  
4     1900-01-01 13:32:50  
...                   ...  
23407 1900-01-01 08:22:12  
23408 1900-01-01 09:13:47  
23409 1900-01-01 12:38:51  
23410 1900-01-01 22:30:19  
23411 1900-01-01 20:08:28  

[23412 rows x 25 columns]

In [78]:
 df.loc[0, 'time']

Timestamp('1900-01-01 13:44:18')

In [84]:
 (df.loc[0, 'time']).time()

datetime.time(13, 44, 18)

Yonidagi yilini olib tashlash qoldi!